gem/optimise.py

```diff
@@ -413,3 +480,565 @@ def aggressive_unroll(expression):
     expression, = unroll_indexsum((expression,), max_extent=numpy.inf)
     expression, = remove_componenttensors((expression,))
     return expression
+
+
+# ---------------------------------------
+# Count flop of expression, "as it is", no reordering etc
+@singledispatch
+def _count_flop(node, self):
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+@_count_flop.register(IndexSum)
+def _count_flop_single(node, self):
+    return self(node.children[0])
+
+
+@_count_flop.register(MathFunction)
+@_count_flop.register(LogicalNot)
+@_count_flop.register(LogicalAnd)
+@_count_flop.register(LogicalOr)
+def _count_flop_func(node, self):
+    return self(node.children[0]) + 1
+
+
+@_count_flop.register(Conditional)  # this is not quite right
+@_count_flop.register(Power)
+@_count_flop.register(Comparison)
+@_count_flop.register(Sum)
+@_count_flop.register(Product)
+@_count_flop.register(Division)
+def _count_flop_common(node, self):
+    flop = numpy.product([i.extent for i in node.free_indices])
+    # Hoisting the factors
+    for child in node.children:
+        flop += self(child)
+    return flop
+
+
+@_count_flop.register(Constant)
+@_count_flop.register(Terminal)
+@_count_flop.register(Indexed)
+@_count_flop.register(FlexiblyIndexed)
+def _count_flop_const(node, self):
+    return 0
+
+
+def count_flop(node):
+    mapper = Memoizer_cf(_count_flop)
+    return mapper(node)
+
+
+@singledispatch
+def _expand_all_product(node, self, index):
+    # ---------------------------------------
+    # Expand all products recursively if free index of node include index
+    # from :param:`index`
+    # e.g (a+(b+c)d)e = ae + bde + cde
+
+    raise AssertionError("cannot handle type %s" % type(node))
+
+
+_expand_all_product.register(Node)(reuse_if_untouched_arg)
+
+
+@_expand_all_product.register(Product)
+def _expand_all_product_common(node, self, index):
+    a = self(node.children[0], index)
+    b = self(node.children[1], index)
+    if isinstance(b, Sum) and any([i in b.free_indices for i in index]):
```

Comment:
So you are iterating over `index`, meaning that it isn't an index, but rather a list/tuple of indices. So this name is misleading, use `indices` or `multiindex` instead, though I think `indices` is right here.